Add alternate optional flow where users need to create account and login before applying

frjo opened the pull request:
Fixes #3064

gmurtaza00 committed:
Register user succesfull without activation email

gmurtaza00 committed:
Added proper validations for application submission view and register buttons to homepage

gmurtaza00 committed:
Registration successful with email activation link

gmurtaza00 committed:
Hashed password storage and activation email message

gmurtaza00 committed:
Corrected flow after activation and removed unneccessary field from regitser form

gmurtaza00 committed:
Errors messages corrected

gmurtaza00 committed:
Code Cleaned

gmurtaza00 committed:
Password confirmation field added

gmurtaza00 committed:
Register link added on login screen. Redirect to apply after login handled

gmurtaza00 committed:
Registration form in a more django native way

gmurtaza00 committed:
Register link adjusted on login page and login redirection handled.

gmurtaza00 committed:
Restricted application form to logged in users only

Commit:
Fix errors caused by cherry-picking original change

When cherry picking, some errors manifested during the conflict
resolution.  This cleans them up and adds a little whitespace.

Commit:
Use settings variables for enabling registration

Adds the two following variables, and updates the registration code to
look for them to conditionally enable the feature:

* FORCE_LOGIN_FOR_APPLICATION
* ENABLE_REGISTRATION_WITHOUT_APPLICATION

Commit:
Add registration tests

frankduncan commented:
Resolves #3064 

Seeks to add two features:
* Allowing signing up before submitting an application
* Requiring login before submitting an application

This is enabled by two variables:
* ENABLE_REGISTRATION_WITHOUT_APPLICATION for the former
* FORCE_LOGIN_FOR_APPLICATION for the latter

(As a side note, as this is my first PR of hopefully many from OTS's work with a client on enhancing hypha, any generic code and PR practices that I failed to follow would be greatly appreciated)

theskumar commented:
I'm off this week but will take this up next week.

theskumar commented on hypha/apply/templates/forms/includes/field.html:
```diff
@@ -35,26 +37,31 @@
+                </div>
+            {% endif %}
+        </div>
+    {%else%}
```
```suggestion
    {% else %}
```

theskumar commented on hypha/apply/templates/forms/includes/field.html:
```diff
@@ -35,26 +37,31 @@
+        </div>
+    {%else%}
+        {{field}}
+    {%endif%}
```
```suggestion
    {% endif %}
```

theskumar commented on hypha/apply/templates/forms/includes/field.html:
```diff
@@ -35,26 +37,31 @@
-        {% if widget_type == 'date_input' or widget_type == 'date_time_input' %}
-            <div class="{{ widget_type }}">
-        {% endif %}
+    {% if not field.is_hidden%}
```
```suggestion
    {% if not field.is_hidden %}
```

theskumar commented on hypha/apply/templates/forms/includes/field.html:
```diff
@@ -35,26 +37,31 @@
+            {% if field.errors %}<h6 class="form__error-text">{{ field.errors.as_text|linebreaksbr }}</h6>{% endif %}
+            {% if widget_type == 'clearable_file_input' or widget_type == 'multi_file_input' %}
+                <!-- <div class="form__file-drop-zone">Drop files here</div> -->
+                <!-- <h1>{{ widget_type }}</h1> -->
```
You can remove the comments if not needed.

theskumar commented on hypha/apply/templates/forms/includes/field.html:
```diff
@@ -19,12 +19,14 @@
-                <span class="form__required">*</span>
-            {% endif %}
-        </label>
+        {% if not field.is_hidden%}
```
```suggestion
        {% if not field.is_hidden %}
```

theskumar commented on hypha/apply/users/models.py:
```diff
@@ -153,14 +153,20 @@ def get_or_create_and_notify(self, defaults: dict | None =None, site=None, **kwa
+                temp_pass_hash = make_password(kwargs.pop('password'))
+            else:
+                temp_pass = BaseUserManager().make_random_password(length=32)
+                temp_pass_hash = make_password(temp_pass)
```
```suggestion
            if 'password' in kwargs:
                # Coming from registration without application
                temp_pass = kwargs.pop('password')
            else:
                temp_pass = BaseUserManager().make_random_password(length=32)
                
            temp_pass_hash = make_password(temp_pass)
```

This can be simplified.

theskumar commented on hypha/apply/users/views.py:
```diff
@@ -241,7 +278,7 @@ def get(self, request, *args, **kwargs):
             user.backend = settings.CUSTOM_AUTH_BACKEND
             login(request, user)
-            return redirect('users:activate_password')
+            return redirect('users:account')
```
Curious why this change is needed?

frankduncan replied:
This change was based on the workflow that we adopted here, were three things are set:

1. We require login to make an application
2. You can register before making an application (forced by 1.)
3. Local passwords are used, and set during registration (as opposed to using 2FA)

What that means is that upon activating your account, you've already set your password.  So you don't need to do that again.  You should just get redirected to your account page.

I can understand how this change is actually bad when and this change got missed when cherry picking our old internal commits, so it should definitely be removed (since the default behavior should remain unchanged).  But maybe the correct answer is to do some kind of switch depending on whether the user set the password, or it was set to a random one by the system?

What do you think?

theskumar replied:
I think we can put this as conditional under the new setting introduced by you. So it retains the old behavior and updates only when login/register is required before submission. 

@frjo thoughts?

frankduncan replied:
There's also the path where registration is enabled, but passwords aren't part of it (and are either set later or they use 2FA).  So maybe conditional on both of those, otherwise the user gets sent to the activation page.

So both `WAGTAIL_PASSWORD_MANAGEMENT_ENABLED` and `ENABLE_REGISTRATION_WITHOUT_APPLICATION`

theskumar replied:
@frankduncan That sounds good to me. Let's have that in.

frankduncan replied:
All set!  Though it turned out to be `WAGTAILUSERS_PASSWORD_ENABLED`, not `WAGTAIL_PASSWORD_MANAGEMENT_ENABLED` for the password fields to be available during registration

theskumar reviewed (changes requested):
Had a quick looks at the code and it looks solid. 

I'll try it out and test it later next week. In mean time, I've left some suggestions for you to add. 

Also, would be great if you can update the `docs/` with the new configurations. Thanks.

frankduncan commented:
Thanks for the quick review!  Comments to come inline :)  Look forward to more next week.

Commit:
Document the two new settings variables

theskumar commented:
Thanks @frankduncan for making changes, I'll be able to review them tomorrow.

theskumar commented:
I tested the feature locally, and looks great to me! Will be waiting for https://github.com/HyphaApp/hypha/pull/3156#discussion_r1097364115 to get resolved.

Commit:
Make the post account creation redirect setting dependent

In response to the conversation here:

https://github.com/HyphaApp/hypha/pull/3156#discussion_r1097364115

When both registering and wagtail password setting is enabled, the user
will put in their password during the registration process, and
therefore shouldn't be redirected after activation to set their
password.  Instead, just redirect to their account page.

frjo commented:
@frankduncan @theskumar  This works very well in my testing.
